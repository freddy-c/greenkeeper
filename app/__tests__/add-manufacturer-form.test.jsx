import "@testing-library/jest-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import AddManufacturerForm from "../manufacturer/add-manufacturer-form";
import { prismaMock } from "../singleton";

jest.mock("next/cache", () => ({
    __esModule: true,
    revalidatePath: jest.fn(() => null),
}));

describe("AddManufacturerForm", () => {
    test("renders correctly", () => {
        const tree = renderer.create(<AddManufacturerForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("shows error message when name is empty", async () => {
        render(<AddManufacturerForm />);
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => {
            expect(
                screen.getByText(/Required/i)
            ).toBeInTheDocument();
        });
    });

    test("successfully submits form with valid name", async () => {
        const validData = { name: "New Manufacturer" }; // Provide valid data based on your schema
        const mockData = { id: "1", name: "New Manufacturer" };

        prismaMock.manufacturer.findMany.mockResolvedValue([]); // Mock no existing manufacturer
        prismaMock.manufacturer.create.mockResolvedValue(mockData); // Mock successful creation

        render(<AddManufacturerForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() =>
            expect(prismaMock.manufacturer.create).toHaveBeenCalledWith({
                data: {
                    name: "Test Manufacturer",
                },
            })
        );
    });

    test("resets form after successful submission", async () => {
        const mockData = { id: "1", name: "Test Manufacturer" };

        prismaMock.manufacturer.findMany.mockResolvedValue([]);
        prismaMock.manufacturer.create.mockResolvedValue(mockData); // Mock successful creation

        render(<AddManufacturerForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => {
            expect(nameInput.value).toBe("");
        });
    });

    test("shows error message for duplicate name", async () => {
        const mockData = { id: "1", name: "Test Manufacturer" };
        prismaMock.manufacturer.findMany.mockResolvedValue([mockData]);

        render(<AddManufacturerForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => {
            expect(
                screen.getByText(/This name is already in use/i)
            ).toBeInTheDocument();
        });
    });
});
