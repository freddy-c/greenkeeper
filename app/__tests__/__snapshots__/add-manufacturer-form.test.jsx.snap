// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddManufacturerForm renders correctly 1`] = `
<div
  className="space-y-6 max-w-lg mx-auto"
>
  <form
    className="space-y-8"
    onSubmit={[Function]}
  >
    <div
      className="space-y-2"
    >
      <label
        className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        htmlFor=":r0:-form-item"
        onMouseDown={[Function]}
      >
        Name
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid={false}
        aria-label="name"
        className="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
      />
      <p
        className="text-sm text-muted-foreground"
        id=":r0:-form-item-description"
      >
        This is the name of the manufacturer.
      </p>
    </div>
    <button
      className="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      type="submit"
    >
      Add Manufacturer
    </button>
  </form>
</div>
`;
